### PR TITLE
[mlir]Fix dialect conversion drop uses

### PR DESCRIPTION
**Pull request by rsuderman**
Before deleting the block we need to drop uses to the surrounding args. If this is not performed dialect conversion failures can result in a failure to remove args (despite the block having no remaining uses).

### CODE REVIEW

**Comment by joker-eph** [comment]
Can you add a test for this?

File: mlir/lib/Transforms/Utils/DialectConversion.cpp [hunk 1]
```diff
@@ -282,8 +282,10 @@ class CreateBlockRewrite : public BlockRewrite {
     block->dropAllUses();
```
Comment (matthias-springer) [inline mlir/lib/Transforms/Utils/DialectConversion.cpp:282]: You can turn this one into `block->dropAllDefinedValueUses();` instead. `dropAllUses` is called from `dropAllDefinedValueUses`.

**Comment by matthias-springer** [comment]
Interesting, when I refactored the code, I removed some calls to `dropAllDefinedValueUses` etc. because I thought there shouldn't be any uses at the point of rollback. (And no test case failed.) It would be great to have a test case for this.

**Comment by joker-eph** [comment]
First time I see “dc” as a title tag, have we used this before?  We already spell out “dialect conversion” fully anyway in the title.

**Comment by rsuderman** [comment]
> First time I see “dc” as a title tag, have we used this before? We already spell out “dialect conversion” fully anyway in the title.

Yup not a problem. Added it mainly because I am used to doing subtags.

File: mlir/lib/Transforms/Utils/DialectConversion.cpp (hunk 1 above)
Comment (rsuderman) [reply mlir/lib/Transforms/Utils/DialectConversion.cpp:282]: done.

**Comment by rsuderman** [comment]
> Can you add a test for this?

Test added - I had to piggy back on top of the `tosa-to-linalg` pass as it appears related to args -> terminator, so defining that with a custom lowering ends with more work than would be preferred.

File: mlir/lib/Transforms/Utils/DialectConversion.cpp [hunk 2]
```diff
@@ -282,8 +282,10 @@ class CreateBlockRewrite : public BlockRewrite {
...
     if (block->getParent())
       block->erase();
-    else
+    else {
```
Comment (joker-eph) [inline mlir/lib/Transforms/Utils/DialectConversion.cpp:285]: Nit: braces around the if when you need them for the else.

**Review by joker-eph — APPROVED** [review]

File: mlir/lib/Transforms/Utils/DialectConversion.cpp (hunk 2 above)
Comment (rsuderman) [reply mlir/lib/Transforms/Utils/DialectConversion.cpp]: done

**Comment by vitalybuka** [comment]
Could you please fix or revert https://lab.llvm.org/buildbot/#/builders/236/builds/10337 ?

**Pull request merged**